### PR TITLE
[pornhub] extract all formats

### LINKED ISSUE #27386: [PornHub] 240p format not being detected
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2020.12.12**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['http://www.pornhub.com/view_video.php?viewkey=648719015', '--list-formats', '--verbose']
[debug] Encodings: locale cp1252, fs mbcs, out cp1252, pref cp1252
[debug] youtube-dl version 2020.12.12
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.19042
[debug] exe versions: none
[debug] Proxy map: {}
[PornHub] 648719015: Downloading pc webpage
[PornHub] 648719015: Downloading m3u8 information
[PornHub] 648719015: Downloading m3u8 information
[PornHub] 648719015: Downloading m3u8 information
[PornHub] 648719015: Downloading m3u8 information
WARNING: unable to extract like count; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
WARNING: unable to extract dislike count; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
[info] Available formats for 648719015:
format code  extension  resolution note
hls-355      mp4        426x240     355k , avc1.640015, 29.97fps, mp4a.40.2
hls-377      mp4        426x240     377k , avc1.640015, 29.97fps, mp4a.40.2
hls-799-0    mp4        854x480     799k , avc1.64001f, 29.97fps, mp4a.40.2
hls-799-1    mp4        854x480     799k , avc1.64001f, 29.97fps, mp4a.40.2
hls-1500-0   mp4        1280x720   1500k , avc1.64001f, 29.97fps, mp4a.40.2
hls-1500-1   mp4        1280x720   1500k , avc1.64001f, 29.97fps, mp4a.40.2
480p         mp4        480p       2000k 
720p         mp4        720p       4000k  (best)
```

## Description

Something changed on PornHub the past 2 days and the 240p format isn't being detected for all videos I've tried. Also like/dislike count isn't being extracted consistently. Works for some videos and doesn't work for others (like the one used above, or this random one from the home page: https://www.pornhub.com/view_video.php?viewkey=ph5ea1a27312cf8).

### PR DESCRIPTION
**Pull request by JChris246**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

As indicated by #27386 some formats were neglected. This PR seeks to rectify this issue by extracting the urls from the json qualityItems which holds the missing 240p quality along with other alternative formats.

closes #27386

### CODE REVIEW

File: youtube_dl/extractor/pornhub.py [hunk 1]
```diff
@@ -281,20 +281,31 @@ def parse_js_value(inp):
 
         def add_video_url(video_url):
             v_url = url_or_none(video_url)
-            if not v_url:
+            if not v_url or not isinstance(v_url, compat_str):
```
Comment (dstftw) [inline youtube_dl/extractor/pornhub.py:284]: Superfluous.

File: youtube_dl/extractor/pornhub.py [hunk 2]
```diff
@@ -281,20 +281,31 @@ def parse_js_value(inp):
...
                 return
             if v_url in video_urls_set:
                 return
             video_urls.append((v_url, None))
             video_urls_set.add(v_url)
 
+        def parse_quality_items(js_str):
+            media_definitions = self._parse_json(js_str, None)
```
Comment (dstftw) [inline youtube_dl/extractor/pornhub.py:292]: `None` is not a video id.

File: youtube_dl/extractor/pornhub.py [hunk 3]
```diff
@@ -281,20 +281,31 @@ def parse_js_value(inp):
...
+                    add_video_url(video_url)
+
         if not video_urls:
             FORMAT_PREFIXES = ('media', 'quality')
             js_vars = extract_js_vars(
                 webpage, r'(var\s+(?:%s)_.+)' % '|'.join(FORMAT_PREFIXES),
                 default=None)
             if js_vars:
                 for key, format_url in js_vars.items():
+                    if key.startswith("qualityItems"):
```
Comment (dstftw) [inline youtube_dl/extractor/pornhub.py:307]: Single quotes.

File: youtube_dl/extractor/pornhub.py (hunk 2 above)
Comment (dstftw) [inline youtube_dl/extractor/pornhub.py:292]: Breaks.

File: youtube_dl/extractor/pornhub.py [hunk 4]
```diff
@@ -281,20 +281,31 @@ def parse_js_value(inp):
...
             video_urls.append((v_url, None))
             video_urls_set.add(v_url)
 
+        def parse_quality_items(js_str):
+            media_definitions = self._parse_json(js_str, None)
+            if isinstance(media_definitions, list):
+                for definition in media_definitions:
+                    if not isinstance(definition, dict):
+                        continue
+                    video_url = definition.get('url')
```
Comment (dstftw) [inline youtube_dl/extractor/pornhub.py:297]: Inline.

**Review by dstftw — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/pornhub.py (hunk 4 above)
Comment (JChris246) [reply youtube_dl/extractor/pornhub.py:297]: ??

Comment (dstftw) [reply youtube_dl/extractor/pornhub.py:297]: ?

Comment (JChris246) [reply youtube_dl/extractor/pornhub.py:297]: What do you mean by inline. Am I to change the for loop to a list comprehension??

Comment (dstftw) [reply youtube_dl/extractor/pornhub.py:297]: Inline the variable to the place of usage. What's not clear?

File: youtube_dl/extractor/pornhub.py [hunk 5]
```diff
@@ -288,13 +288,23 @@ def add_video_url(video_url):
...
+                    add_video_url(definition.get('url'))
+
         if not video_urls:
             FORMAT_PREFIXES = ('media', 'quality')
             js_vars = extract_js_vars(
                 webpage, r'(var\s+(?:%s)_.+)' % '|'.join(FORMAT_PREFIXES),
                 default=None)
             if js_vars:
                 for key, format_url in js_vars.items():
+                    if key.startswith('qualityItems'):
```
Comment (dstftw) [inline youtube_dl/extractor/pornhub.py:306]: Regex must look for this kind of prefix.

**Review by dstftw — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/pornhub.py [hunk 6]
```diff
@@ -297,16 +297,17 @@ def parse_quality_items(js_str):
...
                 default=None)
             if js_vars:
                 for key, format_url in js_vars.items():
-                    if key.startswith('qualityItems'):
-                        parse_quality_items(format_url)
                     if any(key.startswith(p) for p in FORMAT_PREFIXES):
-                        add_video_url(format_url)
+                        if key.startswith('qualityItems'):
+                            parse_quality_items(format_url)
+                        else:
```
Comment (dstftw) [inline youtube_dl/extractor/pornhub.py:309]: This is not equivalent, not matching `qualityItems` does not mean it will match `media` or `quality`.

**Review by dstftw — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/pornhub.py [hunk 7]
```diff
@@ -288,13 +288,23 @@ def add_video_url(video_url):
...
-            FORMAT_PREFIXES = ('media', 'quality')
+            FORMAT_PREFIXES = ('media', 'quality', 'qualityItems')
             js_vars = extract_js_vars(
                 webpage, r'(var\s+(?:%s)_.+)' % '|'.join(FORMAT_PREFIXES),
                 default=None)
             if js_vars:
                 for key, format_url in js_vars.items():
+                    if key.startswith('qualityItems'):
+                        parse_quality_items(format_url)
                     if any(key.startswith(p) for p in FORMAT_PREFIXES):
```
Comment (dstftw) [inline youtube_dl/extractor/pornhub.py:308]: You've already checked for `qualityItems` two lines above.

**Review by dstftw — CHANGES_REQUESTED** [review]

**Pull request closed without merging**